### PR TITLE
[SPARK-30733][R][HOTFIX] Fix SparkR tests per testthat and R version upgrade, and disable CRAN

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

There are currently the R test failures after upgrading `testthat` to 2.0.0, and R version 3.5.2 as of SPARK-23435. This PR targets to fix the tests and make the tests pass. See the explanations and causes below:

```
test_context.R:49: failure: Check masked functions
length(maskedCompletely) not equal to length(namesOfMaskedCompletely).
1/1 mismatches
[1] 6 - 4 == 2

test_context.R:53: failure: Check masked functions
sort(maskedCompletely, na.last = TRUE) not equal to sort(namesOfMaskedCompletely, na.last = TRUE).
5/6 mismatches
x[2]: "endsWith"
y[2]: "filter"

x[3]: "filter"
y[3]: "not"

x[4]: "not"
y[4]: "sample"

x[5]: "sample"
y[5]: NA

x[6]: "startsWith"
y[6]: NA
```

From my cursory look, R base and R's version are mismatched. I fixed accordingly and Jenkins will test it out.

```
test_includePackage.R:31: error: include inside function
package or namespace load failed for ���plyr���:
 package ���plyr��� was installed by an R version with different internals; it needs to be reinstalled for use with this R version
Seems it's a package installation issue. Looks like plyr has to be re-installed.
```

From my cursory look, previously installed `plyr` remains and it's not compatible with the new R version. I fixed accordingly and Jenkins will test it out.

```
test_sparkSQL.R:499: warning: SPARK-17811: can create DataFrame containing NA as date and time
Your system is mis-configured: ���/etc/localtime��� is not a symlink
```

Seems a env problem. I suppressed the warnings for now.

```
test_sparkSQL.R:499: warning: SPARK-17811: can create DataFrame containing NA as date and time
It is strongly recommended to set envionment variable TZ to ���America/Los_Angeles��� (or equivalent)
```

Seems a env problem. I suppressed the warnings for now.

```
test_sparkSQL.R:1814: error: string operators
unable to find an inherited method for function ���startsWith��� for signature ���"character"���
1: expect_true(startsWith("Hello World", "Hello")) at /home/jenkins/workspace/SparkPullRequestBuilder@2/R/pkg/tests/fulltests/test_sparkSQL.R:1814
2: quasi_label(enquo(object), label)
3: eval_bare(get_expr(quo), get_env(quo))
4: startsWith("Hello World", "Hello")
5: (function (classes, fdef, mtable) 
   {
       methods <- .findInheritedMethods(classes, fdef, mtable)
       if (length(methods) == 1L) 
           return(methods[[1L]])
       else if (length(methods) == 0L) {
           cnames <- paste0("\"", vapply(classes, as.character, ""), "\"", collapse = ", ")
           stop(gettextf("unable to find an inherited method for function %s for signature %s", 
               sQuote(fdef@generic), sQuote(cnames)), domain = NA)
       }
       else stop("Internal error in finding inherited methods; didn't return a unique method", 
           domain = NA)
   })(list("character"), new("nonstandardGenericFunction", .Data = function (x, prefix) 
   {
       standardGeneric("startsWith")
   }, generic = structure("startsWith", package = "SparkR"), package = "SparkR", group = list(), 
       valueClass = character(0), signature = c("x", "prefix"), default = NULL, skeleton = (function (x, 
           prefix) 
       stop("invalid call in method dispatch to 'startsWith' (no default method)", domain = NA))(x, 
           prefix)), <environment>)
6: stop(gettextf("unable to find an inherited method for function %s for signature %s", 
       sQuote(fdef@generic), sQuote(cnames)), domain = NA)
```

From my cursory look, R base and R's version are mismatched. I fixed accordingly and Jenkins will test it out.

Also, this PR causes a CRAN check failure as below:

```
* creating vignettes ... ERROR
Error: processing vignette 'sparkr-vignettes.Rmd' failed with diagnostics:
package ���htmltools��� was installed by an R version with different internals; it needs to be reinstalled for use with this R version
```

This PR disables it for now.

### Why are the changes needed?

To unblock other PRs.

### Does this PR introduce any user-facing change?

No. Test only and dev only.

### How was this patch tested?

No. I am going to use Jenkins to test.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Please let me test this via Jenkins considering the urgency.

**Comment by HyukjinKwon** [comment]
cc @dongjoon-hyun and @viirya

**Comment by viirya** [comment]
Will look into it later today after dinner.

**Comment by dongjoon-hyun** [comment]
cc @felixcheung and @shivaram , too.

**Comment by shivaram** [comment]
Thanks @HyukjinKwon -- Changes LGTM if Jenkins tests pass

**Comment by dongjoon-hyun** [comment]
So far, it looks good except the following. I'm not sure why `arrow` test coverage is still removed. Previously, IIRC, we have this.
```
test_sparkSQL_arrow.R:25: skip: createDataFrame/collect Arrow optimization
arrow cannot be loaded
```

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. For Arrow, we can check later.
Since this PR is a HOTFIX, please merge this first.

**Comment by dongjoon-hyun** [comment]
Thank you so much for quick fix, @HyukjinKwon !

**Comment by dongjoon-hyun** [comment]
Oh, it seems that there are another instances.
```
test_sparkSQL.R:504: warning: SPARK-17811: can create DataFrame containing NA as date and time
Your system is mis-configured: ���/etc/localtime��� is not a symlink

test_sparkSQL.R:504: warning: SPARK-17811: can create DataFrame containing NA as date and time
It is strongly recommended to set envionment variable TZ to ���America/Los_Angeles��� (or equivalent)
```

**Review by viirya — APPROVED** [review]
Thanks @HyukjinKwon for quick fix!

**Comment by HyukjinKwon** [comment]
Thanks all @dongjoon-hyun, @viirya, @shivaram. I am sure now all instances are fixed. I will merge as soon as the tests pass.

**Comment by HyukjinKwon** [comment]
Okay, the tests passed and warnings all are fixed.

```
Duration: 598.4 s

OK:       2270
Failed:   0
Warnings: 0
Skipped:  11
```

I think it's a script problem. let me take a look

**Comment by dongjoon-hyun** [comment]
Yes, now it's blocked by `vignettes` error.
```
* creating vignettes ... ERROR
Error: processing vignette 'sparkr-vignettes.Rmd' failed with diagnostics:
package ���htmltools��� was installed by an R version with different internals; it needs to be reinstalled for use with this R version
```

**Comment by HyukjinKwon** [comment]
Argh ..

File: R/create-rd.sh
Comment (HyukjinKwon) [inline R/create-rd.sh]: This should be less critical since the SparkR tests themselves pass. I will summarize the summary for followup actions after I make it pass and merge.

File: R/run-tests.sh
Comment (dongjoon-hyun) [inline R/run-tests.sh]: This reminds me the following. :)
- 91d990162f ([SPARK-24152][R][TESTS] Disable check-cran from run-tests.sh)

Comment (dongjoon-hyun) [reply R/run-tests.sh]: You need remove the following line `FAILED=$((PIPESTATUS[0]||$FAILED))` together.

Comment (HyukjinKwon) [reply R/run-tests.sh]: Thanks @dongjoon-hyun

**Comment by HyukjinKwon** [comment]
I am going to merge to unblock other PRs.

**Comment by HyukjinKwon** [comment]
Merged to master, branch-3.0, and branch-2.4.

**Comment by HyukjinKwon** [comment]
@shaneknapp to summarize what happen so far:

- Seems R version itself bumped up correctly to 3.5.2; however, seems we should reinstall the packages installed previously. It caused some problems with error messages such as:

  ```
  * creating vignettes ... ERROR
  Error: processing vignette 'sparkr-vignettes.Rmd' failed with diagnostics:
  package ���htmltools��� was installed by an R version with different internals; it needs to be 
  reinstalled for use with this R version
  ```

  Currently, I fixed it by making tests permissive and just skipping CRAN check but we should at least reenable CRAN back. Can we reinstall the packages previously installed for R 3.5.2?

  Once this is done, I should reenable it [here](https://github.com/apache/spark/pull/27460/files#diff-9e63dce390dd4d80c6abe3413a8b1984R34) per SPARK-30737

- Looks like Arrow R library was unable to find after the upgrade:

  ```
  test_sparkSQL_arrow.R:25: skip: createDataFrame/collect Arrow optimization
  arrow cannot be loaded
  ```

  This seems causing to skip Arrow related tests in Jenkins.

- I think this is minor but seems there are some enviornment issues assuming from:

  ```
  test_sparkSQL.R:499: warning: SPARK-17811: can create DataFrame containing NA as date and time
  Your system is mis-configured: ���/etc/localtime��� is not a symlink
  ```
  ```
  test_sparkSQL.R:499: warning: SPARK-17811: can create DataFrame containing NA as date and time
  It is strongly recommended to set envionment variable TZ to ���America/Los_Angeles��� (or 
  equivalent)
  ```

  Once this is done, I should remove [this line](https://github.com/apache/spark/pull/27460/files#diff-d4011863c8b176830365b2f224a84bf2R504). But this isn't really a big deal.

**Comment by HyukjinKwon** [comment]
ah cc @zero323 too.

**Comment by zero323** [comment]
Thanks for working on this @HyukjinKwon!

Side question ‒ is `testthat` on Jenkins really 2.0. not 2.3?

**Comment by shaneknapp** [comment]
> @shaneknapp to summarize what happen so far:
> 
> * Seems R version itself bumped up correctly to 3.5.2; however, seems we should reinstall the packages installed previously, including `r-base`. It caused some problems with error messages such as the below (please refer the PR description).

yeah, upgrading `testthat` lead to my worst nightmare:  the game of R dependency whack-a-mole.

i'm currently continuing playing right now and will continue to update packages i find.

>   Currently, I fixed it by making tests permissive and just skipping CRAN check but we should at least reenable CRAN back. Can we reinstall the packages previously installed?  It should be able to test via manually calling `./R/check-cran.sh` or reverting the changes made in this PR at `R/run-tests.sh `.

sounds good.  i'm testing manually on a centos worker and will update this PR when i think i've got everything.

> * Looks like Arrow R library was unable to find after the upgrade:
>   ```
>   test_sparkSQL_arrow.R:25: skip: createDataFrame/collect Arrow optimization
>   arrow cannot be loaded
>   ```
>   
interesting.  i'll install that too.

> * I think this is minor but seems there are some enviornment issues assuming from:
>   ```
>   test_sparkSQL.R:499: warning: SPARK-17811: can create DataFrame containing NA as date and time
>   Your system is mis-configured: ���/etc/localtime��� is not a symlink
>   ```
>   
that is definitely odd.  will investigate.

have i mentioned how much i hate touching R?  :)

**Comment by shaneknapp** [comment]
btw debugging happening here:  https://github.com/apache/spark/pull/27468

it appears that i'll need to open up a new PR (once testing is done) for this commit:  https://github.com/apache/spark/pull/27468/commits/ecbc087873c06b83ca563c162abed2f6f0bf7ba1

**Comment by shaneknapp** [comment]
> Thanks for working on this @HyukjinKwon!
> 
> Side question ‒ is `testthat` on Jenkins really 2.0. not 2.3?

yes, 2.0.0:

`testthat     "testthat"     "/usr/lib64/R/library" "2.0.0"      NA`

**Comment by zero323** [comment]
@shaneknapp 

> yes, 2.0.0:
> 
> `testthat "testthat" "/usr/lib64/R/library" "2.0.0" NA`

I know I shouldn't ask, but is there any chance to get it up to 2.3? 2.0 is already pretty old, and considering its entanglement with tidyverse it will sooner or later block something. And it would be nice to have it synced with Windows builds.

**Comment by shaneknapp** [comment]
> @shaneknapp
> 
> > yes, 2.0.0:
> > `testthat "testthat" "/usr/lib64/R/library" "2.0.0" NA`
> 
> I know I shouldn't ask, but is there any chance to get it up to 2.3? 2.0 is already pretty old, and considering its entanglement with tidyverse it will sooner or later block something. And it would be nice to have it synced with Windows builds.

as evidenced by the amount of time it took me to get the build system fixed, i'd rather wait until we absolutely MUST upgrade testthat to 2.3.x.

**Comment by shaneknapp** [comment]
i mean, i'll do it, but not until next week.  after the past few hours of playing dependency whack-a-mole, i think i need a stiff drink.  ;)

**Comment by HyukjinKwon** [comment]
Yeah, from my experience, touching R was really painful. So I tried to make the tests pass first to decouple the tests vs env issue to earn some enough time :-). Thanks, @shaneknapp.

**Comment by shaneknapp** [comment]
and thanks @HyukjinKwon for taking care of this last night.

**Comment by HyukjinKwon** [comment]
I opened a PR enable back at https://github.com/apache/spark/pull/27472

**Pull request closed without merging**
